cloudevents/sdk/http_events.py

```diff
@@ -11,183 +11,148 @@
 #    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 #    License for the specific language governing permissions and limitations
 #    under the License.
-import copy
-
-import io
 
+import copy
+import datetime
 import json
+import io
 import typing
+import uuid
 
 from cloudevents.sdk import converters
 from cloudevents.sdk import marshaller
+from cloudevents.sdk import types
+from cloudevents.sdk.converters import binary
+from cloudevents.sdk.converters import structured
 
 from cloudevents.sdk.event import v03, v1
 
 
 class CloudEvent():
     """
     Python-friendly cloudevent class supporting v1 events
-    Currently only supports binary content mode CloudEvents
+    Supports both binary and structured mode CloudEvents
     """
 
-    def __init__(
-            self,
-            data: typing.Union[dict, None],
-            headers: dict = {},
-            data_unmarshaller: typing.Callable = lambda x: x,
+    @classmethod
+    def FromHttp(
+        cls,
+        data: typing.Union[str, bytes],
+        headers: dict = {},
+        data_unmarshaller: types.UnmarshallerType = None
     ):
+        """Unwrap a CloudEvent (binary or structured) from an HTTP request.
+        :param data: the HTTP request body
+        :type data: typing.IO
+        :param headers: the HTTP headers
+        :type headers: dict
+        :param data_unmarshaller: A function to decode data into a python object.
+        :type data_unmarshaller: types.UnmarshallerType
         """
-        Event HTTP Constructor
-        :param data: a nullable dict to be stored inside Event.
-        :type data: dict or None
-        :param headers: a dict with HTTP headers
+        def json_or_string(content: typing.Union[str, bytes]):
+            if len(content) == 0:
+                return None
+            try:
+                return json.loads(content)
+            except json.JSONDecodeError:
+                return content
+        if data_unmarshaller is None:
+            data_unmarshaller = json_or_string
+
+        event = marshaller.NewDefaultHTTPMarshaller().FromRequest(
+            v1.Event(), headers, data, data_unmarshaller)
+        attrs = event.Properties()
+        attrs.pop('data', None)
+        attrs.pop('extensions', None)
+        attrs.update(**event.extensions)
+
+        return cls(attrs, event.data)
+
+    def __init__(self, attributes: dict = {}, data: typing.Any = None):
```

Comment:
Can we make sure to do "complete type hints":
```suggestion
    def __init__(self, attributes: typing.Dict[str, typing.Any] = None, data: typing.Any = None):
```